engine/client/drivers/docker.go

```diff
@@ -194,6 +141,27 @@ func (d *dockerDriver) create(ctx context.Context, imageRef string, opts *Driver
 	})
 }
 
+func resolveImageID(ctx context.Context, imageRef string) (string, error) {
+	ref, err := name.ParseReference(imageRef)
+	if err != nil {
+		return "", errors.Wrap(err, "parsing image reference")
+	}
+	if digest, ok := ref.(name.Digest); ok {
+		// We already have the digest as part of the image ref
+		_, id, ok := strings.Cut(digest.DigestStr(), "sha256:")
+		if !ok {
+			return "", errors.Errorf("invalid image reference %q", imageRef)
+		}
+		return id[:hashLen], nil
+	}
+	if tag, ok := ref.(name.Tag); ok {
+		// Otherwise, fallback to the image tag
+		return tag.TagStr(), nil
+	}
+
+	return "", errors.Errorf("invalid image reference %q", imageRef)
```

Comment:
:thinking: this shouldn't error out - theoretically, you *can* do `docker-image://registry.dagger.io/engine` without a tag. In this case, we should default the `Tag` to `latest`?

tangent: Also what, `name.ParseReference` is weird. Can it not parse mixed tags+digests :thinking: which is totally valid as a thing.

Reply:
Done for fallback to `latest` -- PTAL

re: `name.ParseReference` it seems to be working in all cases:

```
# Tag only
$ _EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-image://registry.dagger.io/engine:v0.12.5 ./bin/dagger functions
      ✔ exec docker run --name dagger-engine-v0.12.5 -d --restart always -v /var/lib/dagger --privileged registry.dagger.io/engine:v0.12.5 --debug 0.2s

# Tag+digest
$  _EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-image://registry.dagger.io/engine:v0.12.5@sha256:7d4613636ea6099e3e99c1063d9bbebece1a3465fffac18b98040aa371a00c74 ./bin/dagger functions
      ✔ exec docker run --name dagger-engine-7d4613636ea6099e -d --restart always -v /var/lib/dagger --privileged registry.dagger.io/engine:v0.12.5@sha256:7d4613636ea6099e3e99c1063d9bbebece1a3465fffac18b98040aa371a00c74 --debug 0.2s

# Digest-only
$  _EXPERIMENTAL_DAGGER_RUNNER_HOST=docker-image://registry.dagger.io/engine@sha256:7d4613636ea6099e3e99c1063d9bbebece1a3465fffac18b98040aa371a00c74 ./bin/dagger functions
      ✔ exec docker run --name dagger-engine-7d4613636ea6099e -d --restart always -v /var/lib/dagger --privileged registry.dagger.io/engine@sha256:7d4613636ea6099e3e99c1063d9bbebece1a3465fffac18b98040aa371a00c74 --debug 0.2s
```